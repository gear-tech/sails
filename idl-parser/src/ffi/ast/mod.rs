use crate::ast;
use std::{
    ffi::{c_char, CString},
    slice, str,
};

pub mod visitor;

#[repr(C)]
pub struct ParseResult {
    program: *mut Program,
    error: Error,
}

#[repr(C)]
pub struct Error {
    code: ErrorCode,
    details: *const c_char,
}

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ErrorCode {
    Ok,
    InvalidIDL,
    ParseError,
    NullPtr,
}

/// # Safety
///
/// See the safety documentation of [`slice::from_raw_parts`].
#[no_mangle]
pub unsafe extern "C" fn parse_idl(idl_ptr: *const u8, idl_len: u32) -> *mut ParseResult {
    let idl_slice = unsafe { slice::from_raw_parts(idl_ptr, idl_len as usize) };
    let idl_str = match str::from_utf8(idl_slice) {
        Ok(s) => s,
        Err(e) => return create_parse_error(ErrorCode::ParseError, e, "validate IDL string"),
    };

    let program = match ast::parse_idl(idl_str) {
        Ok(p) => p,
        Err(e) => return create_parse_error(ErrorCode::ParseError, e, "parse IDL"),
    };

    let result = ParseResult {
        error: Error {
            code: ErrorCode::Ok,
            details: std::ptr::null(),
        },
        program: Box::into_raw(Box::new(program)),
    };

    Box::into_raw(Box::new(result))
}

fn create_parse_error(
    code: ErrorCode,
    e: impl std::error::Error,
    context: &'static str,
) -> *mut ParseResult {
    let details = CString::new(format!("{}: {}", context, e)).unwrap();
    let result = ParseResult {
        error: Error {
            code,
            details: details.into_raw(),
        },
        program: std::ptr::null_mut(),
    };
    Box::into_raw(Box::new(result))
}

/// # Safety
///
/// Pointer must be obtained from [`parse_idl`].
#[no_mangle]
pub unsafe extern "C" fn free_parse_result(result: *mut ParseResult) {
    if result.is_null() {
        return;
    }
    unsafe {
        let result = Box::from_raw(result);
        if result.error.code != ErrorCode::Ok {
            let details = CString::from_raw(result.error.details as *mut i8);
            drop(details);
        }
    }
}

pub type Program = ast::Program;

#[repr(C)]
pub struct Ctor {
    raw_ptr: Ptr,
}

#[repr(C)]
pub struct CtorFunc {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(C)]
pub struct Service {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(C)]
pub struct ServiceFunc {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
    is_query: bool,
}

pub type ServiceEvent = EnumVariant;

#[repr(C)]
pub struct FuncParam {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(C)]
pub struct Type {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(C)]
pub struct TypeDecl {
    raw_ptr: Ptr,
}

pub type PrimitiveType = ast::PrimitiveType;

#[repr(C)]
pub struct StructDef {
    raw_ptr: Ptr,
}

#[repr(C)]
pub struct StructField {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(C)]
pub struct EnumDef {
    raw_ptr: Ptr,
}

#[repr(C)]
pub struct EnumVariant {
    raw_ptr: Ptr,
    name_ptr: *const u8,
    name_len: u32,
}

#[repr(transparent)]
pub struct Ptr(*const ());

impl<T> From<&T> for Ptr {
    fn from(t: &T) -> Self {
        Self(t as *const T as *const ())
    }
}

impl<T> AsRef<T> for Ptr {
    fn as_ref(&self) -> &T {
        unsafe { (self.0 as *const T).as_ref() }.unwrap()
    }
}
