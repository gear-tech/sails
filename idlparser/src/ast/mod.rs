use crate::{grammar::ProgramParser, lexer::Lexer};

pub mod visitor;

pub fn parse_idl(idl: &str) -> Result<Program, String> {
    let lexer = Lexer::new(idl);
    let parser = ProgramParser::new();
    let program = parser.parse(lexer).map_err(|e| format!("{:?}", e))?;
    Ok(program)
}

#[derive(Debug, PartialEq, Clone)]
pub struct Program {
    service: Service,
    types: Vec<Type>,
}

impl Program {
    pub(crate) fn new(service: Service, types: Vec<Type>) -> Self {
        Self { service, types }
    }

    pub fn service(&self) -> &Service {
        &self.service
    }

    pub fn types(&self) -> &[Type] {
        &self.types
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Service {
    funcs: Vec<Func>,
}

impl Service {
    pub(crate) fn new(funcs: Vec<Func>) -> Self {
        Self { funcs }
    }

    pub fn funcs(&self) -> &[Func] {
        &self.funcs
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Func {
    name: String,
    params: Vec<FuncParam>,
    output: TypeDecl,
    is_query: bool,
}

impl Func {
    pub(crate) fn new(
        name: String,
        params: Vec<FuncParam>,
        output: TypeDecl,
        is_query: bool,
    ) -> Self {
        Self {
            name,
            params,
            output,
            is_query,
        }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn params(&self) -> &[FuncParam] {
        &self.params
    }

    pub fn output(&self) -> &TypeDecl {
        &self.output
    }

    pub fn is_query(&self) -> bool {
        self.is_query
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct FuncParam {
    name: String,
    type_decl: TypeDecl,
}

impl FuncParam {
    pub(crate) fn new(name: String, type_decl: TypeDecl) -> Self {
        Self { name, type_decl }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn type_decl(&self) -> &TypeDecl {
        &self.type_decl
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Type {
    name: String,
    def: TypeDef,
}

impl Type {
    pub(crate) fn new(name: String, def: TypeDef) -> Self {
        Self { name, def }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn def(&self) -> &TypeDef {
        &self.def
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum TypeDecl {
    Vector(Box<TypeDecl>),
    Array {
        item: Box<TypeDecl>,
        len: u32,
    },
    Map {
        key: Box<TypeDecl>,
        value: Box<TypeDecl>,
    },
    Optional(Box<TypeDecl>),
    Result {
        ok: Box<TypeDecl>,
        err: Box<TypeDecl>,
    },
    Id(TypeId),
    Def(TypeDef),
}

#[derive(Debug, PartialEq, Clone)]
pub enum TypeId {
    Primitive(PrimitiveType),
    UserDefined(String),
}

#[derive(Debug, PartialEq, Clone, Copy)]
#[repr(u8)]
pub enum PrimitiveType {
    Null,
    Bool,
    Char,
    Str,
    U8,
    U16,
    U32,
    U64,
    U128,
    I8,
    I16,
    I32,
    I64,
    I128,
}

impl PrimitiveType {
    pub(crate) fn str_to_enum(str: &str) -> Option<Self> {
        match str {
            "bool" => Some(PrimitiveType::Bool),
            "char" => Some(PrimitiveType::Char),
            "str" => Some(PrimitiveType::Str),
            "u8" => Some(PrimitiveType::U8),
            "u16" => Some(PrimitiveType::U16),
            "u32" => Some(PrimitiveType::U32),
            "u64" => Some(PrimitiveType::U64),
            "u128" => Some(PrimitiveType::U128),
            "i8" => Some(PrimitiveType::I8),
            "i16" => Some(PrimitiveType::I16),
            "i32" => Some(PrimitiveType::I32),
            "i64" => Some(PrimitiveType::I64),
            "i128" => Some(PrimitiveType::I128),
            _ => None,
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum TypeDef {
    Struct(StructDef),
    Enum(EnumDef),
}

#[derive(Debug, PartialEq, Clone)]
pub struct StructDef {
    fields: Vec<StructField>,
}

impl StructDef {
    pub(crate) fn new(fields: Vec<StructField>) -> Self {
        Self { fields }
    }

    pub fn fields(&self) -> &[StructField] {
        &self.fields
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct StructField {
    name: Option<String>,
    type_decl: TypeDecl,
}

impl StructField {
    pub(crate) fn new(name: Option<String>, type_decl: TypeDecl) -> Self {
        Self { name, type_decl }
    }

    pub fn name(&self) -> Option<&str> {
        self.name.as_deref()
    }

    pub fn type_decl(&self) -> &TypeDecl {
        &self.type_decl
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct EnumDef {
    variants: Vec<EnumVariant>,
}

impl EnumDef {
    pub(crate) fn new(variants: Vec<EnumVariant>) -> Self {
        Self { variants }
    }

    pub fn variants(&self) -> &[EnumVariant] {
        &self.variants
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct EnumVariant {
    name: String,
    type_decl: Option<TypeDecl>,
}

impl EnumVariant {
    pub(crate) fn new(name: String, type_decl: Option<TypeDecl>) -> Self {
        Self { name, type_decl }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn type_decl(&self) -> Option<&TypeDecl> {
        self.type_decl.as_ref()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parser_works() {
        let program_idl = r"
          type ThisThatSvcAppTupleStruct = struct {
            bool,
          };

          type ThisThatSvcAppDoThatParam = struct {
            p1: u32,
            p2: str,
            p3: ThisThatSvcAppManyVariants,
          };

          type ThisThatSvcAppManyVariants = enum {
            One,
            Two: u32,
            Three: opt u32,
            Four: struct { a: u32, b: opt u16 },
            Five: struct { str, u32 },
            Six: struct { u32 },
            Seven: [map (u32, str), 10],
          };

          service {
            DoThis : (p1: u32, p2: str, p3: struct { opt str, u8 }, p4: ThisThatSvcAppTupleStruct) -> struct { str, u32 };
            DoThat : (param: ThisThatSvcAppDoThatParam) -> result (struct { str, u32 }, struct { str });
            query This : (v1: vec u16) -> u32;
            query That : (v1: null) -> result (str, str);
          };

          type T = enum { One }
        ";

        let program = parse_idl(program_idl).unwrap();

        assert_eq!(program.types().len(), 4);

        //println!("ast: {:#?}", program);
    }

    #[test]
    fn parser_requires_service() {
        let program_idl = r"
          type T = enum { One };
        ";

        let program = parse_idl(program_idl);

        assert!(program.is_err());
    }

    #[test]
    fn parser_requires_single_service() {
        let program_idl = r"
          service {};
          service {}
        ";

        let program = parse_idl(program_idl);

        assert!(program.is_err());
    }

    #[test]
    fn parser_accepts_types_service() {
        let program_idl = r"
          type T = enum { One };
          service {}
        ";

        let program = parse_idl(program_idl).unwrap();

        assert_eq!(program.types().len(), 1);
    }

    #[test]
    fn parser_requires_semicolon_between_types_and_service() {
        let program_idl = r"
          type T = enum { One }
          service {}
        ";

        let program = parse_idl(program_idl);

        assert!(program.is_err());
    }

    #[test]
    fn parser_accepts_service_types() {
        let program_idl = r"
          service {};
          type T = enum { One };
        ";

        let program = parse_idl(program_idl).unwrap();

        assert_eq!(program.types().len(), 1);
    }

    #[test]
    fn parser_requires_semicolon_between_service_and_types() {
        let program_idl = r"
          service {}
          type T = enum { One };
        ";

        let program = parse_idl(program_idl);

        assert!(program.is_err());
    }
}
