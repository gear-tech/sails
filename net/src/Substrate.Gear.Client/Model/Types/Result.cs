using Substrate.NetApi.Model.Types;
using Substrate.NetApi.Model.Types.Base;

namespace Substrate.Gear.Client.Model.Types;

/// <summary>
/// Result
/// </summary>
public enum Result
{

    /// <summary>
    /// >> Ok
    /// </summary>
    Ok = 0,

    /// <summary>
    /// >> Err
    /// </summary>
    Err = 1,
}

/// <summary>
/// EnumResult
/// </summary>
public sealed class EnumResult<T1, T2> : BaseEnumRust<Result>
    where T1 : IType, new() where T2 : IType, new()
{

    /// <summary>
    /// Initializes a new instance of the class.
    /// </summary>
    public EnumResult()
    {
        this.AddTypeDecoder<T1>(Result.Ok);
        this.AddTypeDecoder<T2>(Result.Err);
    }
}
