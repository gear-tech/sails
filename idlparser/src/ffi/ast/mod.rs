use crate::ast;
use std::{slice, str};

pub mod visitor;

/// # Safety
///
/// See the safety documentation of [`slice::from_raw_parts`].
#[no_mangle]
pub unsafe extern "C" fn parse_idl(idl_ptr: *const u8, idl_len: u32) -> *mut Program {
    let idl = unsafe { slice::from_raw_parts(idl_ptr, idl_len.try_into().unwrap()) };
    let idl = str::from_utf8(idl).unwrap();
    let program = ast::parse_idl(idl).unwrap();
    let program = Box::new(program);
    Box::into_raw(program)
}

/// # Safety
///
/// TODO
#[no_mangle]
pub unsafe extern "C" fn free_program(program: *mut Program) {
    if program.is_null() {
        return;
    }
    unsafe {
        drop(Box::from_raw(program));
    }
}

pub type Program = ast::Program;

pub type Service = ast::Service;

#[repr(C, packed)]
pub struct Func {
    name_ptr: *const u8,
    name_len: u32,
    is_query: bool,
    raw_func: *const ast::Func,
}

#[repr(C, packed)]
pub struct FuncParam {
    name_ptr: *const u8,
    name_len: u32,
    raw_func_param: *const ast::FuncParam,
}

#[repr(C, packed)]
pub struct Type {
    name_ptr: *const u8,
    name_len: u32,
    raw_type: *const ast::Type,
}

pub type TypeDecl = ast::TypeDecl;

pub type PrimitiveType = ast::PrimitiveType;

pub type StructDef = ast::StructDef;

#[repr(C, packed)]
pub struct StructField {
    name_ptr: *const u8,
    name_len: u32,
    raw_struct_field: *const ast::StructField,
}

pub type EnumDef = ast::EnumDef;

#[repr(C, packed)]
pub struct EnumVariant {
    name_ptr: *const u8,
    name_len: u32,
    raw_enum_variant: *const ast::EnumVariant,
}
