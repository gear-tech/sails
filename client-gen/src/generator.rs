use anyhow::Result;
use convert_case::{Case, Casing};
use parity_scale_codec::Encode;
use sails_idlparser::ast::*;
use sails_idlparser::{ast::visitor, ast::visitor::Visitor};
use std::io::Write;

pub fn generate(program: Program, service_name: &str) -> Result<String> {
    let mut generator = RootGenerator::new(service_name);
    visitor::accept_program(&program, &mut generator);

    let code = generator.code;

    // Check for parsing errors
    let code = pretty_with_rustfmt(&code);

    Ok(code)
}

// not using prettyplease since it's bad at reporting syntax errors and also removes comments
// TODO(holykol): Fallback if rustfmt is not in PATH would be nice
fn pretty_with_rustfmt(code: &str) -> String {
    use std::process::Command;
    let mut child = Command::new("rustfmt")
        .arg("--config")
        .arg("format_strings=false")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .spawn()
        .expect("Failed to spawn rustfmt");

    let child_stdin = child.stdin.as_mut().expect("Failed to open stdin");
    child_stdin
        .write_all(code.as_bytes())
        .expect("Failed to write to rustfmt");

    let output = child
        .wait_with_output()
        .expect("Failed to wait for rustfmt");

    if !output.status.success() {
        panic!(
            "rustfmt failed with status: {}\n{}",
            output.status,
            String::from_utf8(output.stderr).expect("Failed to read rustfmt stderr")
        );
    }

    String::from_utf8(output.stdout).expect("Failed to read rustfmt output")
}

struct RootGenerator<'a> {
    code: String,
    traits_code: String,
    service_name: &'a str,
}

impl<'a> RootGenerator<'a> {
    fn new(service_name: &'a str) -> Self {
        let mut code = String::new();

        code.push_str("// Code generated by sails-client-gen. DO NOT EDIT.\n");
        code.push_str("use parity_scale_codec::{Encode, Decode};\n");
        code.push_str("#[allow(unused_imports)]\n");
        code.push_str(
            "use sails_rtl::{prelude::*, String, calls::{Call, Activation, Remoting, RemotingAction}};\n",
        );
        code.push_str("#[allow(unused_imports)]\n");
        code.push_str("use sails_rtl::collections::BTreeMap;\n");
        code.push_str("use core::marker::PhantomData;\n");

        Self {
            service_name,
            traits_code: String::new(),
            code,
        }
    }
}

impl<'a, 'ast> Visitor<'ast> for RootGenerator<'a> {
    fn visit_ctor(&mut self, ctor: &'ast Ctor) {
        let mut ctor_gen = CtorTraitGenerator::new(self.service_name.to_owned());
        ctor_gen.visit_ctor(ctor);
        self.traits_code.push_str(&ctor_gen.code);

        let mut ctor_gen = CtorFactoryGenerator::new(self.service_name.to_owned());
        ctor_gen.visit_ctor(ctor);
        self.code.push_str(&ctor_gen.code);
    }

    fn visit_service(&mut self, service: &'ast Service) {
        let mut service_gen = ServiceGenerator::new(self.service_name.to_owned());
        service_gen.visit_service(service);

        self.traits_code.push_str(&service_gen.code);

        self.code.push_str("pub mod traits {\n");
        self.code.push_str("use super::*;\n");
        self.code.push_str(&self.traits_code);
        self.code.push_str("}\n");

        let mut client_gen = ClientGenerator::new(self.service_name.to_owned());
        client_gen.visit_service(service);
        self.code.push_str(&client_gen.code);
    }

    fn visit_type(&mut self, t: &'ast Type) {
        let mut type_gen = TypeGenerator::new(t.name());
        type_gen.visit_type(t);
        self.code.push_str(&type_gen.code);
    }
}

struct CtorTraitGenerator {
    service_name: String,
    code: String,
}

impl CtorTraitGenerator {
    fn new(service_name: String) -> Self {
        Self {
            service_name,
            code: String::new(),
        }
    }
}

impl<'ast> Visitor<'ast> for CtorTraitGenerator {
    fn visit_ctor(&mut self, ctor: &'ast Ctor) {
        self.code.push_str(&format!(
            "pub trait {}Factory<A: Default> {{\n",
            self.service_name
        ));
        visitor::accept_ctor(ctor, self);
        self.code.push_str("}\n");
    }

    fn visit_ctor_func(&mut self, func: &'ast CtorFunc) {
        let fn_name = func.name().to_case(Case::Snake);

        if fn_name == "new" {
            // https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self
            self.code.push_str("#[allow(clippy::new_ret_no_self)]\n");
        }
        self.code
            .push_str(&format!("fn {fn_name}(remoting: impl Remoting<A>,"));

        visitor::accept_ctor_func(func, self);

        self.code.push_str(")-> impl Activation<A>;\n");
    }

    fn visit_func_param(&mut self, func_param: &'ast FuncParam) {
        let type_decl_code = generate_type_decl_code(func_param.type_decl());
        self.code
            .push_str(&format!("{}: {},", func_param.name(), type_decl_code));
    }
}

struct CtorFactoryGenerator {
    service_name: String,
    code: String,
}

impl CtorFactoryGenerator {
    fn new(service_name: String) -> Self {
        Self {
            service_name,
            code: String::new(),
        }
    }
}

impl<'ast> Visitor<'ast> for CtorFactoryGenerator {
    fn visit_ctor(&mut self, ctor: &'ast Ctor) {
        self.code
            .push_str(&format!("pub struct {}Factory;\n", self.service_name));
        self.code
            .push_str(&format!("impl {}Factory {{\n", self.service_name));
        self.code.push_str(
            "
            #[allow(unused)]
            pub fn new() -> Self {
                Self
            }
        ",
        );
        self.code.push_str("}\n");

        self.code.push_str(&format!(
            "impl<A: Default> traits::{}Factory<A> for {}Factory {{\n",
            self.service_name, self.service_name,
        ));
        visitor::accept_ctor(ctor, self);
        self.code.push_str("}\n");
    }

    fn visit_ctor_func(&mut self, func: &'ast CtorFunc) {
        let fn_name = func.name();
        let fn_name_snake = fn_name.to_case(Case::Snake);

        self.code
            .push_str(&format!("fn {fn_name_snake}(remoting: impl Remoting<A>,",));

        visitor::accept_ctor_func(func, self);

        self.code.push_str(")-> impl Activation<A> {\n");
        let args = encoded_args(func.params());

        let route_bytes = fn_name
            .encode()
            .into_iter()
            .map(|x| x.to_string())
            .collect::<Vec<_>>()
            .join(",");

        self.code.push_str(&format!(
            r#"RemotingAction::new(remoting, &[{route_bytes}], {args})"#
        ));

        self.code.push_str("}\n");
    }

    fn visit_func_param(&mut self, func_param: &'ast FuncParam) {
        let type_decl_code = generate_type_decl_code(func_param.type_decl());
        self.code
            .push_str(&format!("{}: {},", func_param.name(), type_decl_code));
    }
}

struct ServiceGenerator {
    service_name: String,
    code: String,
}

impl ServiceGenerator {
    fn new(service_name: String) -> Self {
        Self {
            service_name,
            code: String::new(),
        }
    }
}

impl<'ast> Visitor<'ast> for ServiceGenerator {
    fn visit_service(&mut self, service: &'ast Service) {
        self.code.push_str(&format!(
            "pub trait {}<R, TCallArgs> {{\n",
            self.service_name
        ));
        visitor::accept_service(service, self);
        self.code.push_str("}\n");
    }

    fn visit_service_func(&mut self, func: &'ast ServiceFunc) {
        let mutability = if func.is_query() { "" } else { "mut" };
        let name = func.name().to_case(Case::Snake);

        self.code.push_str("#[allow(clippy::type_complexity)]\n");
        self.code
            .push_str(&format!("fn {}(&{} self,", name, mutability));
        visitor::accept_service_func(func, self);
        self.code.push_str(";\n");
    }

    fn visit_func_param(&mut self, func_param: &'ast FuncParam) {
        let type_decl_code = generate_type_decl_code(func_param.type_decl());
        self.code
            .push_str(&format!("{}: {},", func_param.name(), type_decl_code));
    }

    fn visit_func_output(&mut self, func_output: &'ast TypeDecl) {
        let type_decl_code = generate_type_decl_code(func_output);
        self.code
            .push_str(&format!(") -> impl Call<TCallArgs, {}>", type_decl_code));
    }
}

struct ClientGenerator {
    service_name: String,
    code: String,
}

impl ClientGenerator {
    fn new(service_name: String) -> Self {
        Self {
            service_name,
            code: String::new(),
        }
    }
}

impl<'ast> Visitor<'ast> for ClientGenerator {
    fn visit_service(&mut self, service: &'ast Service) {
        let name = &self.service_name;

        self.code.push_str(&format!(
            r#"
            #[derive(Clone)]
            pub struct {name}<R, A> where
                R: Remoting<A>,
                A: Default,
            {{
                remoting: R,
                _phantom: PhantomData<A>,
            }}

            impl<A: Default, R: Remoting<A>> {name}<R, A> {{
                pub fn new(remoting: R) -> Self {{
                    Self {{ remoting, _phantom: PhantomData }}
                }}
            }}

            impl<R, A> traits::{name}<R, A> for {name}<R, A>
            where
                R: Remoting<A> + Clone,
                A: Default,
            {{
        "#
        ));

        visitor::accept_service(service, self);
        self.code.push_str("}\n");
    }

    fn visit_service_func(&mut self, func: &'ast ServiceFunc) {
        let mutability = if func.is_query() { "" } else { "mut" };
        let fn_name = func.name();

        self.code.push_str(&format!(
            "fn {}(&{} self,",
            fn_name.to_case(Case::Snake),
            mutability
        ));

        visitor::accept_service_func(func, self);

        self.code.push_str("{\n");

        let args = encoded_args(func.params());

        let route_bytes = fn_name
            .encode()
            .into_iter()
            .map(|x| x.to_string())
            .collect::<Vec<_>>()
            .join(",");

        self.code.push_str(&format!(
            "RemotingAction::new(self.remoting.clone(), &[{route_bytes}], {args})"
        ));

        self.code.push_str("}\n");
    }

    fn visit_func_param(&mut self, func_param: &'ast FuncParam) {
        let type_decl_code = generate_type_decl_code(func_param.type_decl());
        self.code
            .push_str(&format!("{}: {},", func_param.name(), type_decl_code));
    }

    fn visit_func_output(&mut self, func_output: &'ast TypeDecl) {
        let type_decl_code = generate_type_decl_code(func_output);
        self.code
            .push_str(&format!(") -> impl Call<A, {}>", type_decl_code));
    }
}

fn encoded_args(params: &[FuncParam]) -> String {
    if params.len() == 1 {
        return params[0].name().to_owned();
    }

    let arg_names = params
        .iter()
        .map(|a| a.name())
        .collect::<Vec<_>>()
        .join(", ");

    format!("({arg_names})")
}

struct TypeGenerator<'a> {
    type_name: &'a str,
    code: String,
}

impl<'a> TypeGenerator<'a> {
    fn new(type_name: &'a str) -> Self {
        Self {
            type_name,
            code: String::new(),
        }
    }
}

impl<'a, 'ast> Visitor<'ast> for TypeGenerator<'a> {
    fn visit_struct_def(&mut self, struct_def: &'ast StructDef) {
        let mut struct_def_generator = StructDefGenerator::default();
        struct_def_generator.visit_struct_def(struct_def);

        let semi = if struct_def.fields().iter().all(|f| f.name().is_none()) {
            ";"
        } else {
            ""
        };

        self.code
            .push_str("#[derive(PartialEq, Debug, Encode, Decode)]");
        self.code.push_str(&format!(
            "pub struct {} {} {}",
            self.type_name, struct_def_generator.code, semi
        ));
    }

    fn visit_enum_def(&mut self, enum_def: &'ast EnumDef) {
        let mut enum_def_generator = EnumDefGenerator::default();
        enum_def_generator.visit_enum_def(enum_def);

        self.code
            .push_str("#[derive(PartialEq, Debug, Encode, Decode)]");
        self.code.push_str(&format!(
            "pub enum {} {}",
            self.type_name, &enum_def_generator.code
        ));
    }
}

#[derive(Default)]
struct StructDefGenerator {
    code: String,
}

impl<'ast> Visitor<'ast> for StructDefGenerator {
    fn visit_struct_def(&mut self, struct_def: &'ast StructDef) {
        let is_regular_struct = struct_def.fields().iter().all(|f| f.name().is_some());
        let is_tuple_struct = struct_def.fields().iter().all(|f| f.name().is_none());
        if !is_regular_struct && !is_tuple_struct {
            panic!("Struct must be either regular or tuple");
        }
        if is_regular_struct {
            self.code.push('{');
        } else {
            self.code.push('(');
        }
        visitor::accept_struct_def(struct_def, self);
        if is_regular_struct {
            self.code.push('}');
        } else {
            self.code.push(')');
        }
    }

    fn visit_struct_field(&mut self, struct_field: &'ast StructField) {
        let type_decl_code = generate_type_decl_code(struct_field.type_decl());

        if let Some(field_name) = struct_field.name() {
            self.code
                .push_str(&format!("{}: {},", field_name, type_decl_code));
        } else {
            self.code.push_str(&format!("{},", type_decl_code));
        }
    }
}

#[derive(Default)]
struct EnumDefGenerator {
    code: String,
}

impl<'ast> Visitor<'ast> for EnumDefGenerator {
    fn visit_enum_def(&mut self, enum_def: &'ast EnumDef) {
        self.code.push('{');
        visitor::accept_enum_def(enum_def, self);
        self.code.push('}');
    }

    fn visit_enum_variant(&mut self, enum_variant: &'ast EnumVariant) {
        if let Some(type_decl) = enum_variant.type_decl().as_ref() {
            let type_decl_code = generate_type_decl_code(type_decl);
            if type_decl_code.starts_with('{') {
                self.code
                    .push_str(&format!("{} {},", enum_variant.name(), type_decl_code));
            } else {
                self.code
                    .push_str(&format!("{}({}),", enum_variant.name(), type_decl_code));
            }
        } else {
            self.code.push_str(&format!("{},", enum_variant.name()));
        }
    }
}

fn generate_type_decl_code(type_decl: &TypeDecl) -> String {
    let mut type_decl_generator = TypeDeclGenerator::default();
    visitor::accept_type_decl(type_decl, &mut type_decl_generator);
    type_decl_generator.code
}

#[derive(Default)]
struct TypeDeclGenerator {
    code: String,
}

impl<'ast> Visitor<'ast> for TypeDeclGenerator {
    fn visit_optional_type_decl(&mut self, optional_type_decl: &'ast TypeDecl) {
        self.code.push_str("Option<");
        visitor::accept_type_decl(optional_type_decl, self);
        self.code.push('>');
    }

    fn visit_result_type_decl(
        &mut self,
        ok_type_decl: &'ast TypeDecl,
        err_type_decl: &'ast TypeDecl,
    ) {
        self.code.push_str("Result<");
        visitor::accept_type_decl(ok_type_decl, self);
        self.code.push_str(", ");
        visitor::accept_type_decl(err_type_decl, self);
        self.code.push('>');
    }

    fn visit_vector_type_decl(&mut self, vector_type_decl: &'ast TypeDecl) {
        self.code.push_str("Vec<");
        visitor::accept_type_decl(vector_type_decl, self);
        self.code.push('>');
    }

    fn visit_struct_def(&mut self, struct_def: &'ast StructDef) {
        let mut struct_def_generator = StructDefGenerator::default();
        struct_def_generator.visit_struct_def(struct_def);
        self.code.push_str(&struct_def_generator.code);
    }

    fn visit_primitive_type_id(&mut self, primitive_type_id: PrimitiveType) {
        self.code.push_str(match primitive_type_id {
            PrimitiveType::U8 => "u8",
            PrimitiveType::U16 => "u16",
            PrimitiveType::U32 => "u32",
            PrimitiveType::U64 => "u64",
            PrimitiveType::U128 => "u128",
            PrimitiveType::I8 => "i8",
            PrimitiveType::I16 => "i16",
            PrimitiveType::I32 => "i32",
            PrimitiveType::I64 => "i64",
            PrimitiveType::I128 => "i128",
            PrimitiveType::Bool => "bool",
            PrimitiveType::Str => "String",
            PrimitiveType::Char => "char",
            PrimitiveType::Null => "()",
            PrimitiveType::ActorId => "ActorId",
            PrimitiveType::CodeId => "CodeId",
            PrimitiveType::MessageId => "MessageId",
        });
    }

    fn visit_user_defined_type_id(&mut self, user_defined_type_id: &'ast str) {
        self.code.push_str(user_defined_type_id);
    }

    fn visit_map_type_decl(
        &mut self,
        key_type_decl: &'ast TypeDecl,
        value_type_decl: &'ast TypeDecl,
    ) {
        self.code.push_str("BTreeMap<");
        visitor::accept_type_decl(key_type_decl, self);
        self.code.push_str(", ");
        visitor::accept_type_decl(value_type_decl, self);
        self.code.push('>');
    }

    fn visit_array_type_decl(&mut self, item_type_decl: &'ast TypeDecl, len: u32) {
        self.code.push('[');
        visitor::accept_type_decl(item_type_decl, self);
        self.code.push_str(&format!("; {len}]"));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_works() {
        let idl = r"
            type MyParam = struct {
                f1: u32,
                f2: vec str,
                f3: opt struct { u8, u32 },
            };

            type MyParam2 = enum {
                Variant1,
                Variant2: u32,
                Variant3: struct { u32 },
                Variant4: struct { u8, u32 },
                Variant5: struct { f1: str, f2: vec u8 },
            };

            service {
                DoThis: (p1: u32, p2: MyParam) -> u16;
                DoThat: (p1: struct { u8, u32 }) -> u8;
            };
        ";

        let program = sails_idlparser::ast::parse_idl(idl).expect("parse IDL");

        insta::assert_snapshot!(generate(program, "Basic").unwrap());
    }

    #[test]
    fn test_rmrk_works() {
        let idl = include_str!("../../examples/rmrk/catalog/wasm/rmrk-catalog.idl");

        let program = sails_idlparser::ast::parse_idl(idl).expect("parse IDL");

        insta::assert_snapshot!(generate(program, "RmrkCatalog").unwrap());
    }
}
