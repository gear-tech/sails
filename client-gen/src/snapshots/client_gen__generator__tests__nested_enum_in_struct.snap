---
source: client-gen/src/generator.rs
expression: generator.generate(program).unwrap()
---
// Code generated by sails-client-gen from test. DO NOT EDIT.
#![automatically_derived]
#![allow(unused)]
use parity_scale_codec::{Encode, Decode};
use sails_client::{Sender, Call};
#[derive(PartialEq, Debug, Encode, Decode)]
pub enum P3Variant  {
One,
Two,
} 
#[derive(PartialEq, Debug, Encode, Decode)]
pub struct ThisThatAppDoThatParam  {
p2: String,
p3: P3Variant,
} 


    #[derive(Default, Clone)]
    pub struct Client {
        program_id: [u8; 32],
    }

    impl Client {

    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_program_id(mut self, program_id: impl Into<[u8; 32]>) -> Self {
        self.program_id = program_id.into();
        self
    }


    
}

